### ISSUE TITLE
Bug: [no-unnecessary-condition] False positive when using `lodash.get()`

### ISSUE THREAD

**Issue by ivan-aksamentov** [typescript-eslint/typescript-eslint#9556]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=JYWwDg9gTgLgBAbzgcwKbwL5wGZQiOAcgBsIATAQwGcALQgKHuwFcA7AYxmAlZwggAUASkT04cdjyrxiFAEapiVAFxwASqklQyAHmlRgrZABo4%2Bw8gB8cALyIMYiVPgA3CsWapbKdANkKlU0IaRVJCEQB%2BCKIIhnEAeni4AFEAZQAZQxhVAFVWVk1UKioKKABPJ1YyYC4ed2NHRLhUAA8wTRhUMjhiVGwYAFoaCiqzYDIvCGw4AAMombgIdqgKGGg4NbgFOEhi4DliCtZmYmJFqDg2CexDLoA6RqSBAAEYMvaqdgMwQaLiLPirAgAzYBXYRRK5QGkiqNW4rCE9AcQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
import { get } from 'lodash'

function foo() {
  const obj: Record<string, string> = {}
  const value = get(obj, 'hello') ?? '?'
  // ESLint: Unnecessary conditional,
  // expected left-hand side of `??` operator to be possibly null or undefined.
  // (@typescript-eslint/no-unnecessary-condition)
}

function foo2() {
  const numbers: number[] = []
  const value = get(numbers, 1) ?? 42
  // ESLint: Unnecessary conditional,
  // expected left-hand side of `??` operator to be possibly null or undefined.
  // (@typescript-eslint/no-unnecessary-condition)
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": "warn",
  },
};
```

### Expected Result

No problems should be reported.

### Actual Result

`@typescript-eslint/no-unnecessary-condition` warning is reported

### Additional Info

I am sorry for not being able to provide a more detailed repro due to lack of time, but I believe that this is happening with any configuration and this is a bug in `@typescript-eslint/no-unnecessary-condition` rule.

The `lodash` typings for `get()` are very complex, some of the "overloads" have non-nil return types (the ones that fallback on provided default value), so the linter must be confused:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/5be58a75483be41164b64f7bf57d7d57e1460a93/types/lodash/common/object.d.ts#L1117-L1362

This is an important enough use-case for a popular library, and following the linter advice in this case can lead to runtime errors (as demonstrated in the example code).

If someone have time, please investigate.

**Comment by bradzacher** [comment]
https://typescript-eslint.io/troubleshooting/typed-linting#typescript-eslint-thinks-my-variable-is-never-nullish--is-any--etc-but-that-is-clearly-not-the-case-to-me

The function does not return a nullish type. The rule acts on types.

**Comment by bradzacher** [comment]
Have you checked the types presented in your actual code?
[TS playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgcwKbwL5wGZQiOAcgBsIATAQwGcALQgKHuwFcA7AYxmAlZwggAUASkT04cdjyrwIAIwBWALjgAlVJKhkAPNKjBWyADRxd+5AD44AXkQYxEqfABuFYs1TWU6AXPnHCNKjEpIRC9gD04eLiAHoA-PR2TGyc3LzY-ABMwqLikqzScKzMILKoUFTKxaXlANoAup4N9vmFLm4eNmgwAtVlFcYAjGHikdFw8YlAA)

You'll find that the I'm quite correct here.

**Comment by bradzacher** [comment]
Yes - stop using lodash.get and use optional chaining along with [`noUncheckedIndexedAccess`](https://www.typescriptlang.org/tsconfig/#noUncheckedIndexedAccess)

The lodash types are just bad here.
